File: codegen/output/java/Fish.java
import com.sun.jna.Pointer;

///
class Fish extends Mappable
{
  public Fish(Pointer p)
  {
    super(p);
  }
  boolean validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.fishs.length; i++)
    {
      if(BaseAI.fishs[i].ID == ID)
      {
        ptr = BaseAI.fishs[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }

    //commands

  ///Command a fish to move to a specified position
  boolean move(int x, int y)
  {
    validify();
    return (Client.INSTANCE.fishMove(ptr, x, y) == 0) ? false : true;
  }
  ///Command a fish to pick up some trash at a specified position
  boolean pickUp(int x, int y, int weight)
  {
    validify();
    return (Client.INSTANCE.fishPickUp(ptr, x, y, weight) == 0) ? false : true;
  }
  ///Command a fish to drop some trash at a specified position
  boolean drop(int x, int y, int weight)
  {
    validify();
    return (Client.INSTANCE.fishDrop(ptr, x, y, weight) == 0) ? false : true;
  }
  ///Command a fish to attack another fish at a specified position
  boolean attack(int x, int y)
  {
    validify();
    return (Client.INSTANCE.fishAttack(ptr, x, y) == 0) ? false : true;
  }

    //getters

  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.fishGetId(ptr);
  }
  ///X position of the object
  public int getX()
  {
    validify();
    return Client.INSTANCE.fishGetX(ptr);
  }
  ///Y position of the object
  public int getY()
  {
    validify();
    return Client.INSTANCE.fishGetY(ptr);
  }
  ///The owner of this fish
  public int getOwner()
  {
    validify();
    return Client.INSTANCE.fishGetOwner(ptr);
  }
  ///The type/species of the fish
  public String getSpecies()
  {
    validify();
    return Client.INSTANCE.fishGetSpecies(ptr);
  }
  ///The maximum health of the fish
  public int getMaxHealth()
  {
    validify();
    return Client.INSTANCE.fishGetMaxHealth(ptr);
  }
  ///The current health of the fish
  public int getCurHealth()
  {
    validify();
    return Client.INSTANCE.fishGetCurHealth(ptr);
  }
  ///The maximum number of movements in a turn
  public int getMaxMoves()
  {
    validify();
    return Client.INSTANCE.fishGetMaxMoves(ptr);
  }
  ///The number of movements left
  public int getMovementLeft()
  {
    validify();
    return Client.INSTANCE.fishGetMovementLeft(ptr);
  }
  ///The total weight the fish can carry
  public int getCarryCap()
  {
    validify();
    return Client.INSTANCE.fishGetCarryCap(ptr);
  }
  ///The current amount of weight the fish is carrying
  public int getCarryWeight()
  {
    validify();
    return Client.INSTANCE.fishGetCarryWeight(ptr);
  }
  ///The power of the fish's attack
  public int getAttackPower()
  {
    validify();
    return Client.INSTANCE.fishGetAttackPower(ptr);
  }
  ///The visibleness of the fish
  public int getIsVisible()
  {
    validify();
    return Client.INSTANCE.fishGetIsVisible(ptr);
  }
  ///The number of attacks a fish has left
  public int getAttacksLeft()
  {
    validify();
    return Client.INSTANCE.fishGetAttacksLeft(ptr);
  }

}

File: codegen/output/visclient/Fish.cpp
// -*-c++-*-

#include "Fish.h"
#include "game.h"


namespace client
{

Fish::Fish(_Fish* pointer)
{
    ptr = (void*) pointer;
}

int Fish::id()
{
  return ((_Fish*)ptr)->id;
}

int Fish::x()
{
  return ((_Fish*)ptr)->x;
}

int Fish::y()
{
  return ((_Fish*)ptr)->y;
}

int Fish::owner()
{
  return ((_Fish*)ptr)->owner;
}

char* Fish::species()
{
  return ((_Fish*)ptr)->species;
}

int Fish::maxHealth()
{
  return ((_Fish*)ptr)->maxHealth;
}

int Fish::curHealth()
{
  return ((_Fish*)ptr)->curHealth;
}

int Fish::maxMoves()
{
  return ((_Fish*)ptr)->maxMoves;
}

int Fish::movementLeft()
{
  return ((_Fish*)ptr)->movementLeft;
}

int Fish::carryCap()
{
  return ((_Fish*)ptr)->carryCap;
}

int Fish::carryWeight()
{
  return ((_Fish*)ptr)->carryWeight;
}

int Fish::attackPower()
{
  return ((_Fish*)ptr)->attackPower;
}

int Fish::isVisible()
{
  return ((_Fish*)ptr)->isVisible;
}

int Fish::attacksLeft()
{
  return ((_Fish*)ptr)->attacksLeft;
}


int Fish::move(int x, int y)
{
  return fishMove( (_Fish*)ptr, x, y);
}

int Fish::pickUp(int x, int y, int weight)
{
  return fishPickUp( (_Fish*)ptr, x, y, weight);
}

int Fish::drop(int x, int y, int weight)
{
  return fishDrop( (_Fish*)ptr, x, y, weight);
}

int Fish::attack(int x, int y)
{
  return fishAttack( (_Fish*)ptr, x, y);
}



std::ostream& operator<<(std::ostream& stream,Fish ob)
{
  stream << "id: " << ((_Fish*)ob.ptr)->id  <<'\n';
  stream << "x: " << ((_Fish*)ob.ptr)->x  <<'\n';
  stream << "y: " << ((_Fish*)ob.ptr)->y  <<'\n';
  stream << "owner: " << ((_Fish*)ob.ptr)->owner  <<'\n';
  stream << "species: " << ((_Fish*)ob.ptr)->species  <<'\n';
  stream << "maxHealth: " << ((_Fish*)ob.ptr)->maxHealth  <<'\n';
  stream << "curHealth: " << ((_Fish*)ob.ptr)->curHealth  <<'\n';
  stream << "maxMoves: " << ((_Fish*)ob.ptr)->maxMoves  <<'\n';
  stream << "movementLeft: " << ((_Fish*)ob.ptr)->movementLeft  <<'\n';
  stream << "carryCap: " << ((_Fish*)ob.ptr)->carryCap  <<'\n';
  stream << "carryWeight: " << ((_Fish*)ob.ptr)->carryWeight  <<'\n';
  stream << "attackPower: " << ((_Fish*)ob.ptr)->attackPower  <<'\n';
  stream << "isVisible: " << ((_Fish*)ob.ptr)->isVisible  <<'\n';
  stream << "attacksLeft: " << ((_Fish*)ob.ptr)->attacksLeft  <<'\n';
  return stream;
}

}
